sympy/core/expr.py

```diff
@@ -808,7 +809,15 @@ def _eval_interval(self, x, a, b):
                 if isinstance(B, Limit):
                     raise NotImplementedError("Could not compute limit")
 
-        return B - A
+        value = B - A
+        a = a if a is not None else 0
+        b = b if b is not None else 0
+        singularities = list(solveset(self.cancel().as_numer_denom()[1], x))
+        for s in singularities:
+            if all([S(a).is_number, S(s).is_number, S(b).is_number])\
```

Comment:
What is the purpose of `S`?

It is better to check  a and b before computing the singularities. Instead of `is_number`, `is_comparable` could probably be used. (Even `I.is_number` is True.)